runtime/src/main/scala/akka/grpc/internal/ClientState.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import java.util.concurrent.CompletionStage
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.grpc.GrpcClientSettings
+import akka.stream.{ ActorMaterializer, Materializer }
+import io.grpc.ManagedChannel
+
+import scala.annotation.tailrec
+import scala.concurrent.{ ExecutionContext, Future, Promise }
+import scala.util.Failure
+import scala.compat.java8.FutureConverters._
+
+/**
+ * INTERNAL API
+ * Used from generated code so can't be private.
+ *
+ * Client utilities taking care of Channel reconnection and Channel lifecycle in general.
+ */
+@InternalApi
+final class ClientState(settings: GrpcClientSettings, channelFactory: GrpcClientSettings => InternalChannel)(implicit mat: Materializer, ex: ExecutionContext) {
+
+  def this(settings: GrpcClientSettings)(implicit mat: Materializer, ex: ExecutionContext) =
+    this(settings, s => NettyClientUtils.createChannel(s))
+
+  private val internalChannelRef =
+    new AtomicReference[Option[InternalChannel]](Some(create()))
+  // usually None, it'll have a value when the underlying InternalChannel is closing or closed.
+  private val closing = new AtomicReference[Option[Future[Done]]](None)
+  private val closeDemand: Promise[Done] = Promise[Done]()
+
+  mat match {
+    case m: ActorMaterializer =>
+      m.system.whenTerminated.foreach(_ => close())(ex)
+    case _ =>
+  }
+
+  def withChannel[A](f: Future[ManagedChannel] => A): A =
+    f {
+      internalChannelRef
+        .get()
+        .getOrElse(throw new ClientClosedException)
+        .managedChannel
+    }
+
+  def closedCS(): CompletionStage[Done] = closed().toJava
+  def closeCS(): CompletionStage[Done] = close().toJava
+
+  def closed(): Future[Done] = {
+    // while there's no request to close this RestartingClient, it will continue to restart.
+    // Once there's demand, the `closeDemand` future will redeem flatMapping with the `closing`
+    // future which is a reference to promise of the internalChannel close status.
+    closeDemand.future.flatMap { _ =>
+      // `closeDemand` guards the read access to `closing`
+      closing.get().get
+    }
+  }
+
+  @tailrec
+  def close(): Future[Done] = {
+    val maybeChannel = internalChannelRef.get()
+    maybeChannel match {
+      case Some(channel) =>
+        // invoke `close` on the channel and capture the `channel.done` returned
+        val done = ChannelUtils.close(channel)
+        // set the `closing` to the current `channel.done`
+        closing.compareAndSet(None, Some(done))
+        // notify there's been close demand (see `def closed()` above)
+        closeDemand.trySuccess(Done)
+
+        if (internalChannelRef.compareAndSet(maybeChannel, None)) {
```

Comment:
Here we can lose this cas but we might have won the previous `closing` cas. Quite confusing, I'd try to avoid more than one atomic an represent the states with and adt instead, something like `case class Open(channelref)`, `case class Closing(Future[Done])` and do a single cas on that.